### PR TITLE
[HELM] All roles in one statefulset

### PR DESCRIPTION
**Pull request by everythings-gonna-be-alright**
Issue #433

Description of changes:
If `elasticsearch.data.enabled` or `elasticsearch.client.enabled` set to false, `elasticsearch.master` StatefulSet automatically will be created with `master` + `data` + `client` roles.
Also `es-data-service`  and `es-client-service` will be routed all queries to master StatefulSet.

Test Results:

1) Deploy using the following chart values :

```
elasticsearch:
  master:
    enabled: true
  data:
    enabled: false
  client:
    enabled: false
```

As result we have 3 master nodes with `data` and `ingest` roles:
```
 > curl -uadmin:admin 192.168.1.254:9200/_cat/nodes                                                                             124⏎ 
172.16.139.170 20  99 13 2.67 2.21 1.82 dimr * elasticsearch-opendistro-opendistro-es-master-2
172.16.39.106  56  97 16 0.81 1.95 2.34 dimr - elasticsearch-opendistro-opendistro-es-master-1
172.16.232.209 60 100 13 1.63 1.38 1.21 dimr - elasticsearch-opendistro-opendistro-es-master-0
```

2) Deploy using the following chart values :

```
elasticsearch:
  master:
    enabled: true
  data:
    enabled: true
  client:
    enabled: true
```
As result we have 9 nodes with dedicated roles:
```
 > curl -uadmin:admin 192.168.1.254:9200/_cat/nodes                                                                             124⏎ 
192.168.53.209  46  93  1 2.98 2.45 2.36 mr * elasticsearch-opendistro-opendistro-es-master-1
192.168.197.33  35 100  2 1.50 1.74 1.64 ir - elasticsearch-opendistro-opendistro-es-client-788b48b5f7-glfgw
192.168.53.253  67 100 43 2.98 2.45 2.36 dr - elasticsearch-opendistro-opendistro-es-data-0
192.168.171.209 66 100 51 2.75 1.96 1.69 dr - elasticsearch-opendistro-opendistro-es-data-2
192.168.197.8   24 100  0 1.50 1.74 1.64 mr - elasticsearch-opendistro-opendistro-es-master-0
192.168.197.36  30 100 15 1.50 1.74 1.64 dr - elasticsearch-opendistro-opendistro-es-data-1
192.168.204.137 49 100  5 0.49 0.37 0.37 ir - elasticsearch-opendistro-opendistro-es-client-788b48b5f7-5xc6f
192.168.153.244 39 100  6 0.37 0.35 0.39 ir - elasticsearch-opendistro-opendistro-es-client-788b48b5f7-xzz5j
192.168.171.239 53  97  0 2.75 1.96 1.69 mr - elasticsearch-opendistro-opendistro-es-master-2
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: helm/opendistro-es/templates/elasticsearch/es-service.yaml [hunk 1]
```diff
@@ -35,6 +34,9 @@ spec:
     - name: rca
       port: 9650
   selector:
-    role: client
+    {{- if .Values.elasticsearch.client.enabled }}
+      role: client
+    {{- else }}
+      role: master
+    {{- end }}
```
Comment (peterzhuamazon) [inline helm/opendistro-es/templates/elasticsearch/es-service.yaml:41]: I think we are 2 space off here right?

**Comment by peterzhuamazon** [comment]
Hi @everythings-gonna-be-alright thanks for the contribution, however:
* We think setting `enabled: false` is to truly disable the node, instead of cancel a dedicated pod, but still being eligible in master node.
* We suggest adding additional param/variable such as "dedicated_host: true/false" such that
  * data node is `enabled:true` and has `dedicated_host:true`, then it is the old behavior, data node exist in dedicated pod;
  * data node is `enabled:true` and has `dedicated_host:false`, then it is the the new behavior you want;
  * data node is `enabled:false` then it is the old behavior, data node not available at all.

Please let us know if this is possible, thanks again for your contribution, appreciate any suggestions.

Thanks.

File: helm/opendistro-es/templates/elasticsearch/es-service.yaml (hunk 1 above)
Comment (everythings-gonna-be-alright) [reply helm/opendistro-es/templates/elasticsearch/es-service.yaml:41]: My bad 😬

**Comment by peterzhuamazon** [comment]
Hi @everythings-gonna-be-alright thanks for the changes.
We will take a look at this one and get back to you.
Thanks.

**Comment by peterzhuamazon** [comment]
Hi @everythings-gonna-be-alright we have reviewed the PR.
Sorry to interrupt again but could you change `dedicatedPod` to `dedicatedPod.enabled` so it is more clear and consistent compares to other parameters?
We apologize that we didnt mention this earlier. Sorry about it.

Please let me know if you can make this change.
Thanks.

**Comment by everythings-gonna-be-alright** [comment]
@peterzhuamazon no problem 🙂 
I think that we could also move all client/data pod settings to dedicatedPod section. It look like more qualitative solution.
Like this:
```
  client:
    enabled: true
    dedicatedPod:
      enalbled: true
      replicas: 1
      javaOpts: "-Xms512m -Xmx512m"
      .....
    service:
      .....
    ingress:
      .....
```
But in this case i'm worried about back compatibility with old values.yaml's. Because if someone will decide to up chart version but will forgot update also yaml file, it can create some issues for him.
And also it look more complicated structure than structure in traditional helm chart.
But anyway, what do you think about this idea ?

**Comment by peterzhuamazon** [comment]
Thanks @everythings-gonna-be-alright for the changes, we will review it again and come back to you if any issues surface.
We appreciate your ideas but concern that it would be a big breaking change to work with.
As of now, I think we can keep HELM as it be.

**Review by sreekarjami — APPROVED** [review]

**Review by peterzhuamazon — APPROVED** [review]

**Comment by peterzhuamazon** [comment]
Thanks @everythings-gonna-be-alright the PR is merged to Master Branch now.

**Pull request merged**